File: capture/hysteresis.py
from collections import deque

class HysteresisBuffer:
    """
    OCR 히트 기록을 바탕으로 상태(선곡화면 진입/이탈)와 신뢰도를 계산합니다.
    """
    def __init__(self, history_size: int, on_ratio: float, on_min_samples: int, off_ratio: float, off_min_samples: int):
        self.history_size = max(1, history_size)
        self.on_ratio = on_ratio
        self.on_min_samples = max(1, on_min_samples)
        self.off_ratio = off_ratio
        self.off_min_samples = max(1, off_min_samples)
        
        self.history = deque(maxlen=self.history_size)
        self.is_active = False
        self.is_leaving = False
        self.confidence = 0.0
        
        # 로그 및 외부 접근용
        self.hit_count = 0
        self.sample_count = 0
        self.ratio = 0.0

    def update(self, is_hit: bool) -> tuple[bool, bool, float]:
        self.history.append(is_hit)
        self.sample_count = len(self.history)
        self.hit_count = sum(1 for v in self.history if v)
        self.ratio = (self.hit_count / self.sample_count) if self.sample_count > 0 else 0.0

        # 진입/이탈 상태 판정
        if self.is_active:
            should_turn_off = (
                self.sample_count >= self.off_min_samples
                and self.ratio <= self.off_ratio
            )
            self.is_active = not should_turn_off
        else:
            self.is_active = (
                self.sample_count >= self.on_min_samples
                and self.ratio >= self.on_ratio
            )

        # 이탈 중(fade-out) 감지
        self.is_leaving = False
        if self.is_active and self.sample_count >= 4:
            half = self.sample_count // 2
            history_list = list(self.history)
            first_half_ratio = sum(history_list[:half]) / half
            second_half_ratio = sum(history_list[half:]) / (self.sample_count - half)
            if second_half_ratio < first_half_ratio:
                self.is_leaving = True

        # 신뢰도 계산
        self.confidence = self.ratio * (0.5 if self.is_leaving else 1.0)
        return self.is_active, self.is_leaving, self.confidence

File: rust/overmax_app/src/hysteresis.rs
use std::collections::VecDeque;

#[derive(Clone, Debug)]
pub struct HysteresisBuffer {
    history_size: usize,
    on_ratio: f32,
    on_min_samples: usize,
    off_ratio: f32,
    off_min_samples: usize,
    history: VecDeque<bool>,
    is_active: bool,
    pub is_leaving: bool,
    pub confidence: f32,
    pub hit_count: usize,
    pub sample_count: usize,
    pub ratio: f32,
}

impl HysteresisBuffer {
    pub fn new(
        history_size: usize,
        on_ratio: f32,
        on_min_samples: usize,
        off_ratio: f32,
        off_min_samples: usize,
    ) -> Self {
        Self {
            history_size: history_size.max(1),
            on_ratio,
            on_min_samples: on_min_samples.max(1),
            off_ratio,
            off_min_samples: off_min_samples.max(1),
            history: VecDeque::new(),
            is_active: false,
            is_leaving: false,
            confidence: 0.0,
            hit_count: 0,
            sample_count: 0,
            ratio: 0.0,
        }
    }

    pub fn update(&mut self, is_hit: bool) -> (bool, bool, f32) {
        self.push(is_hit);
        self.update_counts();
        self.update_active_state();
        self.update_leaving_state();
        self.confidence = self.ratio * if self.is_leaving { 0.5 } else { 1.0 };
        (self.is_active, self.is_leaving, self.confidence)
    }

    fn push(&mut self, is_hit: bool) {
        if self.history.len() == self.history_size {
            self.history.pop_front();
        }
        self.history.push_back(is_hit);
    }

    fn update_counts(&mut self) {
        self.sample_count = self.history.len();
        self.hit_count = self.history.iter().filter(|hit| **hit).count();
        self.ratio = if self.sample_count == 0 {
            0.0
        } else {
            self.hit_count as f32 / self.sample_count as f32
        };
    }

    fn update_active_state(&mut self) {
        if self.is_active {
            let should_turn_off =
                self.sample_count >= self.off_min_samples && self.ratio <= self.off_ratio;
            self.is_active = !should_turn_off;
            return;
        }
        self.is_active = self.sample_count >= self.on_min_samples && self.ratio >= self.on_ratio;
    }

    fn update_leaving_state(&mut self) {
        self.is_leaving = false;
        if !self.is_active || self.sample_count < 4 {
            return;
        }

        let half = self.sample_count / 2;
        let first_hits = self.history.iter().take(half).filter(|hit| **hit).count();
        let second_hits = self.history.iter().skip(half).filter(|hit| **hit).count();
        let first_ratio = first_hits as f32 / half as f32;
        let second_ratio = second_hits as f32 / (self.sample_count - half) as f32;
        self.is_leaving = second_ratio < first_ratio;
    }
}

#[cfg(test)]
mod tests {
    use super::HysteresisBuffer;

    #[test]
    fn activates_after_enough_hits() {
        let mut buffer = HysteresisBuffer::new(7, 0.6, 3, 0.35, 7);
        assert!(!buffer.update(true).0);
        assert!(!buffer.update(true).0);
        assert!(buffer.update(false).0);
    }

    #[test]
    fn detects_leaving_when_recent_hits_drop() {
        let mut buffer = HysteresisBuffer::new(6, 0.5, 3, 0.1, 6);
        for hit in [true, true, true, true, false, false] {
            buffer.update(hit);
        }
        assert!(buffer.is_leaving);
        assert_eq!(buffer.confidence, buffer.ratio * 0.5);
    }
}
